# Part 1. Equation of a Slime

How many late days are you using for this assignment?

In [26]:
# Imports section
import pandas as pd
import sklearn as sk
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score, KFold
import numpy as np
from sklearn import svm
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import PolynomialFeatures
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

## 1. Loading the dataset

In [27]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

df = pd.read_csv(r"C:\Users\sagea\OneDrive\CS 383\Assignment 4\science_data_large.csv")
print(df.head(15))
print("\n\n\nThe table information:")
df.info()

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04



The table information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 

## 2. Splitting the dataset

In [28]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X = df.drop(columns=['Size nm^3']) 
y = df['Size nm^3']   
Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, test_size=0.10, train_size=0.90, random_state=42, shuffle=True)
print(Xtrain)
print(ytrain)

     Temperature °C  Mols KCL
716              95       499
351             243       384
936              47       211
256             495       497
635              82       421
..              ...       ...
106             439        40
270             665       245
860             471       620
435             829       259
102             636       388

[900 rows x 2 columns]
716    103064.31430
351    193753.02860
936     21670.02857
256    505027.40000
635     75092.02857
           ...     
106     40433.71429
270    335545.00000
860    600674.85710
435    441286.60000
102    505469.25710
Name: Size nm^3, Length: 900, dtype: float64


## 3. Perform a Linear Regression

In [29]:
# Use sklearn to train a model on the training set
model = LinearRegression( fit_intercept=True, copy_X=True, n_jobs=None, positive=False)
model.fit(Xtrain, ytrain)
# Create a sample datapoint and predict the output of that sample with the trained model


ypred = model.predict(Xtest)

# Report the score (R-squared value)
# For Linear Regression, the score method returns R^2
score = model.score(Xtest, ytest)
print("The score is: ", score)
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_

print(f"Coefficients: {coefficients}")
print(f"Intercept: {intercept}")

The score is:  0.8552472077276096
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340787


Write the linear equation of a slime: (example equation: $E = mc^2$)

Equation for simple $y = 866.14641x_1 + 1032.69507x_2 -409391.47958 $

Report on score and explain meaning: The score is .85 which means that the linear regression model is performing fairly well and the model is fitting the data correctly and predicting the data about 85% correctly.

## 4. Use Cross Validation

In [30]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
kf = KFold(n_splits=5, shuffle=True, random_state=42)
# Use cross_val_score to evaluate the model
cvscores = cross_val_score(model, X, y, cv=kf)

print(cvscores)
# Report on their finding and their significance

[0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]


Write findings here: The findings are all values in the .80-.90 range. The significance shows that the linear regression model is averaging pretty well when it comes to making accurate predictons based on random samples form the data. 

## 5. Using Polynomial Regression

In [31]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
# Polynomial Features
poly = PolynomialFeatures(degree=2)
Xpoly = poly.fit_transform(X)  

model = LinearRegression()

kf = KFold(n_splits=5, shuffle=True, random_state=42)

polyscores = cross_val_score(model, Xpoly, y, cv=kf)

print(f"Cross-validation scores: {polyscores}")
print(f"Average cross-validation score: {polyscores.mean()}")

model.fit(Xpoly, y)


print("coefficients", model.coef_)
print("intercept:", model.intercept_)
# Report on the metrics and output the resultant equation as you did in Part 3.

Cross-validation scores: [1. 1. 1. 1. 1.]
Average cross-validation score: 1.0
coefficients [ 0.00000000e+00  1.20000000e+01 -1.23112504e-07 -1.05671555e-11
  2.00000000e+00  2.85714287e-02]
intercept: 1.6571604646742344e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$$ y = 1.20000000(e+01)x_2 -1.23112504(e-07)x_3 -1.05671555(e-11)x_4 + 2x_5 +   2.85714287(e-02)x_6 + 1.6571604646742344(e-05) $$

Report on the score and interpret: The score is 1 which means that the polynomial is performing really really well. The intercept is a number very close to 0 ($1.6571604646742344*10^{-5}$)

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [32]:
# Load the dataset. Then train and evaluate the classification models.

ckddf = pd.read_csv(r"C:\Users\sagea\OneDrive\CS 383\Assignment 4\ckd_feature_subset.csv")

print(ckddf.head())
Xckd = ckddf.drop(columns=['Target_ckd'])
yckd = ckddf['Target_ckd']  

Xc_train, Xc_test, yc_train, yc_test = train_test_split(Xckd, yckd, test_size=0.1, train_size=.90, random_state=42)

        age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0  0.688312  0.333333  0.000000           1           0  0.000000           1
1  0.545455  0.333333  0.128319           1           0  0.305085           1
2  0.714286  0.500000  0.238938           1           0  0.186441           1
3  0.688312  0.333333  0.283186           0           1  0.338983           1
4  0.441558  0.333333  0.221239           1           0  0.220339           1


## Results and Conclusion for Classification Experiments

### Logistic Regression:

In [33]:
LGmodel = LogisticRegression()
LGmodel.fit(Xc_train, yc_train)

gkf = KFold(n_splits=5, shuffle=True, random_state=42)
# Use cross_val_score to evaluate the model

LGcvscores = cross_val_score(LGmodel, Xckd, yckd, cv=gkf)
LGstddev = np.std(LGcvscores)
LGmean = np.mean(LGcvscores)
print(f"The cross validation scores for the logistic regression are: {LGcvscores}")
print(f"The standard deviation for the logistic regression CV scores is: {LGstddev}")
print(f"The the mean for the logistic regression CV scores is: {LGmean}")
# Report on their finding and their significance

The cross validation scores for the logistic regression are: [0.80645161 0.93548387 0.77419355 0.83333333 0.93333333]
The standard deviation for the logistic regression CV scores is: 0.0662686697755876
The the mean for the logistic regression CV scores is: 0.8565591397849461


### Support Vector Models

In [34]:
SVMmodel = svm.SVC(kernel='linear')
SVMmodel.fit(Xc_train, yc_train)
gkf = KFold(n_splits=5, shuffle=True, random_state=42)

SVMcvscores = cross_val_score(SVMmodel, Xckd, yckd, cv=gkf)
SVMstddev = np.std(SVMcvscores)
SVMmean = np.mean(SVMcvscores)
print(f"The cross validation scores for the SVM are: {SVMcvscores}")
print(f"The standard deviation for the SVM CV scores is: {SVMstddev}")
print(f"The the mean for the SVM CV scores is: {SVMmean}")
# Report on their finding and their significance

The cross validation scores for the SVM are: [0.80645161 0.96774194 0.77419355 0.83333333 0.93333333]
The standard deviation for the SVM CV scores is: 0.07467538723169277
The the mean for the SVM CV scores is: 0.8630107526881721


### K-Nearest Neighbors

In [35]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(Xc_train)
X_test_scaled = scaler.transform(Xc_test)
knn = KNeighborsClassifier(n_neighbors=5) # Choose k=5 (or another appropriate value)
knn.fit(X_train_scaled, yc_train)

gkf = KFold(n_splits=5, shuffle=True, random_state=42)

knncvscores = cross_val_score(knn, Xckd, yckd, cv=gkf)
knnstddev = np.std(knncvscores)
knnmean = np.mean(knncvscores)
print(f"The cross validation scores for the KNN are: {knncvscores}")
print(f"The standard deviation for the KNN CV scores is: {knnstddev}")
print(f"The the mean for the KNN CV scores is: {knnmean}")
# Report on their finding and their significance

The cross validation scores for the KNN are: [0.93548387 1.         0.87096774 0.86666667 0.96666667]
The standard deviation for the KNN CV scores is: 0.05243961560473918
The the mean for the KNN CV scores is: 0.9279569892473118


In [36]:
NNmodel = MLPClassifier(hidden_layer_sizes=(10,), max_iter=2000, solver='adam', random_state=42)

# Train the model
NNmodel.fit(Xc_train, yc_train)

# Predict on the test set

gkf = KFold(n_splits=5, shuffle=True, random_state=42)
NNcvscores = cross_val_score(NNmodel, Xckd, yckd, cv=gkf)
NNstddev = np.std(NNcvscores)
NNmean = np.mean(NNcvscores)
print(f"The cross validation scores for the Neural Network are: {NNcvscores}")
print(f"The standard deviation for the Network CV scores is: {NNstddev}")
print(f"The the mean for the Neural Network CV scores is: {NNmean}\n")

The cross validation scores for the Neural Network are: [0.90322581 1.         0.83870968 0.9        0.96666667]
The standard deviation for the Network CV scores is: 0.05630713719142086
The the mean for the Neural Network CV scores is: 0.9217204301075268



#### Markdown Code: 

| Model/Configuration        | Cross Validation Scores                                      | Standard Deviation | Mean Cross Validation Score |
|----------------------------|------------------------------------------------------------|--------------------|-----------------------------|
| **Logistic Regression**     | [0.80645161, 0.93548387, 0.77419355, 0.83333333, 0.93333333] | 0.06627            | 0.85656                     |
| **SVM**                    | [0.80645161, 0.96774194, 0.77419355, 0.83333333, 0.93333333] | 0.07468            | 0.86301                     |
| **KNN**                    | [0.93548387, 1.0, 0.87096774, 0.86666667, 0.96666667]        | 0.05244            | 0.92796                     |
| **Neural Network**         | [0.90322581, 1.0, 0.83870968, 0.9, 0.96666667]               | 0.05631            | 0.92172                     |

### Analysis

I believe the K nearest neighbors Model performed the best because it had the highest mean cross-validation score and the lowest standard deviation meaning it performs the most consistently with the highest score. This is probably because the data is more suited for a more flexible model like KNN as opposed to Logistic Regression and SVM, though the other models performed fairly well as well. Upon reinspection and talking to Professor McNichols, I added the extra row for the Neural Network at the bottom of the table. It outperformed the Logistic Regression and SVM, however was only slightly beaten out by the KNN. This upholds my guesses for why the KNN is best suited for the data. 

## Results and Conclusion for Neural Network Experiments

In [37]:
### Control Group:
yc_pred = NNmodel.predict(Xc_test)

# Evaluate the model
accuracy = accuracy_score(yc_test, yc_pred)

print(f'Control Test accuracy: {accuracy}')
print(f"The cross validation scores for the Control are: {NNcvscores}")
print(f"The standard deviation for the Control CV scores is: {NNstddev}")
print(f"The the mean for the Control CV scores is: {NNmean}\n")
# Report on their finding and their significance

### Increased Hidden Layer_sizes
NNmodel = MLPClassifier(hidden_layer_sizes=(30,), max_iter=2000, solver='adam', random_state=42)

# Train the model
NNmodel.fit(Xc_train, yc_train)

# Predict on the test set
yc_pred = NNmodel.predict(Xc_test)

# Evaluate the model
accuracy = accuracy_score(yc_test, yc_pred)
print(f'Increased Hidden Layer Sizes Test accuracy: {accuracy}')

NNcvscores = cross_val_score(NNmodel, Xckd, yckd, cv=gkf)
NNstddev = np.std(NNcvscores)
NNmean = np.mean(NNcvscores)
print(f"The cross validation scores for the Increased Hidden Layer Sizes are: {NNcvscores}")
print(f"The standard deviation for the Increased Hidden Layer Sizes CV scores is: {NNstddev}")
print(f"The the mean for the Increased Hidden Layer Sizes CV scores is: {NNmean}\n")
# Report on their finding and their significance

### Different Hidden Layer_sizes
NNmodel = MLPClassifier(hidden_layer_sizes=(15,5), max_iter=2000, solver='adam', random_state=42)

# Train the model
NNmodel.fit(Xc_train, yc_train)

# Predict on the test set
yc_pred = NNmodel.predict(Xc_test)

# Evaluate the model
accuracy = accuracy_score(yc_test, yc_pred)
print(f'Different Hidden Layer Sizes Test accuracy: {accuracy}')

NNcvscores = cross_val_score(NNmodel, Xckd, yckd, cv=gkf)
NNstddev = np.std(NNcvscores)
NNmean = np.mean(NNcvscores)
print(f"The cross validation scores for the Different Hidden Layer Sizes are: {NNcvscores}")
print(f"The standard deviation for the Different Hidden Layer Sizes CV scores is: {NNstddev}")
print(f"The the mean for the Different Hidden Layer Sizes CV scores is: {NNmean}\n")
# Report on their finding and their significance


### Increased Double Hidden Layer_sizes
NNmodel = MLPClassifier(hidden_layer_sizes=(20,20), max_iter=2000, solver='adam', random_state=42)

# Train the model
NNmodel.fit(Xc_train, yc_train)

# Predict on the test set
yc_pred = NNmodel.predict(Xc_test)

# Evaluate the model
accuracy = accuracy_score(yc_test, yc_pred)
print(f'Increased Double Hidden Layer_sizes Test accuracy: {accuracy}')

NNcvscores = cross_val_score(NNmodel, Xckd, yckd, cv=gkf)
NNstddev = np.std(NNcvscores)
NNmean = np.mean(NNcvscores)
print(f"The cross validation scores for the Increased Double Hidden Layer_sizes are: {NNcvscores}")
print(f"The standard deviation for the Increased Double Hidden Layer_sizes CV scores is: {NNstddev}")
print(f"The the mean for the Increased Double Hidden Layer_sizes CV scores is: {NNmean}\n")
# Report on their finding and their significance

### Different Solver
NNmodel = MLPClassifier(hidden_layer_sizes=(10,), max_iter=2000, solver='lbfgs', random_state=42)

# Train the model
NNmodel.fit(Xc_train, yc_train)

# Predict on the test set
yc_pred = NNmodel.predict(Xc_test)

# Evaluate the model
accuracy = accuracy_score(yc_test, yc_pred)
print(f'lbfgs Test accuracy: {accuracy}')

NNcvscores = cross_val_score(NNmodel, Xckd, yckd, cv=gkf)
NNstddev = np.std(NNcvscores)
NNmean = np.mean(NNcvscores)
print(f"The cross validation scores for the lbfgs solver are: {NNcvscores}")
print(f"The standard deviation for the lbfgs solver CV scores is: {NNstddev}")
print(f"The the mean for the lbfgs solver  CV scores is: {NNmean}\n")
# Report on their finding and their significance

### Different Solver
NNmodel = MLPClassifier(hidden_layer_sizes=(10,), max_iter=2000, solver='sgd', random_state=42)

# Train the model
NNmodel.fit(Xc_train, yc_train)

# Predict on the test set
yc_pred = NNmodel.predict(Xc_test)

# Evaluate the model
accuracy = accuracy_score(yc_test, yc_pred)
print(f'sgd Test accuracy: {accuracy}')

NNcvscores = cross_val_score(NNmodel, Xckd, yckd, cv=gkf)
NNstddev = np.std(NNcvscores)
NNmean = np.mean(NNcvscores)
print(f"The cross validation scores for the sgd solver are: {NNcvscores}")
print(f"The standard deviation for the sgd solver CV scores is: {NNstddev}")
print(f"The the mean for the sgd solver CV scores is: {NNmean}\n")
# Report on their finding and their significance

Control Test accuracy: 0.875
The cross validation scores for the Control are: [0.90322581 1.         0.83870968 0.9        0.96666667]
The standard deviation for the Control CV scores is: 0.05630713719142086
The the mean for the Control CV scores is: 0.9217204301075268

Increased Hidden Layer Sizes Test accuracy: 0.8125
The cross validation scores for the Increased Hidden Layer Sizes are: [0.93548387 1.         0.83870968 0.96666667 0.96666667]
The standard deviation for the Increased Hidden Layer Sizes CV scores is: 0.05530102404560463
The the mean for the Increased Hidden Layer Sizes CV scores is: 0.9415053763440862

Different Hidden Layer Sizes Test accuracy: 0.875
The cross validation scores for the Different Hidden Layer Sizes are: [0.90322581 1.         0.87096774 1.         0.96666667]
The standard deviation for the Different Hidden Layer Sizes CV scores is: 0.052335444281831296
The the mean for the Different Hidden Layer Sizes CV scores is: 0.9481720430107528

Increased Double 

### Findings

| Model/Configuration                          | Test Accuracy | Cross Validation Scores                                      | Standard Deviation   | Mean Cross Validation Score |
|----------------------------------------------|---------------|--------------------------------------------------------------|-----------------------|-----------------------------|
| **Control**                                  | 0.875         | [0.90322581, 1.0, 0.83870968, 0.9, 0.96666667]               | 0.05630713719142086   | 0.9217204301075268          |
| **Increased Hidden Layer Sizes**             | 0.8125        | [0.93548387, 1.0, 0.83870968, 0.96666667, 0.96666667]         | 0.05530102404560463   | 0.9415053763440862          |
| **Different Hidden Layer Sizes**            | 0.875         | [0.90322581, 1.0, 0.87096774, 1.0, 0.96666667]               | 0.052335444281831296  | 0.9481720430107528          |
| **Increased Double Hidden Layer Sizes**     | 0.9375        | [0.93548387, 1.0, 0.87096774, 1.0, 0.96666667]               | 0.04822385846151442   | 0.9546236559139786          |
| **lbfgs Solver**                             | 0.9375        | [1.0, 1.0, 0.90322581, 1.0, 0.96666667]                      | 0.03765836411156565   | 0.9739784946236559          |
| **sgd Solver**                               | 0.6875        | [0.80645161, 0.93548387, 0.77419355, 0.83333333, 0.9]         | 0.05954543187873574   | 0.8498924731182795          |

### Summary

I chose a control group using the adam solver and the parameter for the hidden layers set to (10,_)... I found that as I increased the number of hidden layers to (30,_), the test accuracy initially decreased, but the mean CV scores increased. Next as I found that you can have two different numbers of neurons in hidden layers when I marginally increased the number of neurons in both hidden layers (by 5) so the amount of hidden layers was (15,5) the mean CV scores slightly increased. Finally, I saw another more incremental increase when I increased both hidden layers to (20,20). These incremental improvements in CV accuracy show that as the number of neurons in hidden layers increases so does the accuracy of the model. 

Next, I also texted the different solvers only changing the parameters of the solver with the same number of hidden layers for each model: (10,_). Since the control group was adam, it had a mean CV score of .92. I found that the lbfgs solver had the highest accuracy with a mean CV score of .97, and I found that the sgd had a MUCH lower Test Accuracy and mean CV score accuracy of .6875 and .85 respectively. Based on these results the lbfgs solver is by far the best and performs better than adam even with increased hidden layers, and the sgd solver should be avoided because it had the lowest accuracy. 